pmd-core/src/main/java/net/sourceforge/pmd/reporting/RuleContext.java

```diff
@@ -130,22 +134,43 @@ public void addViolationWithMessage(Node location, String message, Object... for
      * @param formatArgs Format arguments for the message
      */
     public void addViolationWithPosition(Node node, int beginLine, int endLine, String message, Object... formatArgs) {
-        Objects.requireNonNull(node, "Node was null");
+        FileLocation location;
+        if (beginLine != -1 && endLine != -1) {
+            location = FileLocation.range(node.getTextDocument().getFileId(),
+                                          TextRange2d.range2d(beginLine, 1, endLine, 1));
+        } else {
+            location = node.getReportLocation();
+        }
+        addViolationWithPosition(node, node.getAstInfo(), location, message, formatArgs);
+    }
+
+    /**
+     * Record a new violation of the contextual rule, at the given location (node or token).
+     * The position is refined using the given begin and end line numbers.
+     * The given violation message ({@link Rule#getMessage()}) is treated
+     * as a format string for a {@link MessageFormat} and should hence use
+     * appropriate escapes. The given formatting arguments are used.
+     *
+     * @param reportable Location of the violation (node or token)
+     * @param astInfo    Info about the root of the tree ({@link Node#getAstInfo()})
+     * @param message    Violation message
+     * @param formatArgs Format arguments for the message
+     */
+    public void addViolationWithPosition(Reportable reportable, AstInfo<?> astInfo, FileLocation location,
+                                         String message, Object... formatArgs) {
```

Comment:
One problem I have with this is the unreadability of these overloads... Maybe this API could have a makeover? Maybe we should write it like that:
```java
ctx.report(at(node), fmt("msg", a, b))
ctx.report(at(token, astInfo, loc), msg, a, b)

// or

ctx.at(node).warnWithDefaultMsg();
ctx.at(node).warnWithDefaultMsg(arg1, arg2);
ctx.at(token, astInfo, loc).warnWithDefaultMsg();
ctx.at(token, astInfo, loc).warnWithDefaultMsg();
ctx.at(token, astInfo, loc).warn("Custom msg {0}", arg);
```

We could also probably dispense with the `token` and `astInfo` parameters with some magic (store the AstInfo in the RuleContext) and get
```java
ctx.at(loc).warn(...);
```
But apparently changing the construction of the RuleContext would require changing a bunch of test code... Just a small hurdle though if this is what we want.

@jsotuyod @adangel any thoughts?

Reply:
Maybe we could implement it with a builder? Which could look like (similar to your suggestions):

```java
ctx.newViolation(message).withArgs(arg1, arg2).at(node);
ctx.newViolation(message).withArgs(arg1, arg2).at(token, astInfo);
```

In this example, `at(Reportable)` would actually add the ruleviolation to the rule context. We could also have a separate, explicit build method, which would allow to refine the location:

```java
ctx.newViolation(message).withArgs(arg1, arg2).at(node).atBeginLine(4).build();
```

This could be an additional API first, as an alternative to the overloads addViolation...

We could do it slightly different:

```java
new ViolationMessageBuilder().with(message).withArgs(arg).at(node).build(ctx);
```

Which would be then more separated from RuleContext, but honestly, I like it that the entry point for reporting violations is in rulecontext.

Reply:
Late to the party, but here are some thoughts:

1. I support the fluent API, I think it's a neat solution. Moreover, it's very flexible looking forward for adding things like #4327)
2. I feel shaping up that in this PR would be muddying the waters / delay getting this fixed, I'd recommend a separate PR.
3. Any fluent API needs to be flagged `@Experimental`, I'd not commit our first iteration to a stable public API.
4. Starting at `RuleContext` is the cleanest approach, I like it. This is irrespective of wether that is passed down directly or still obtained through `asCtx(data)` (again, separate issue / PR eventually).
5. I'd have the finalizer method be a verb (not `at`!), descriptive (ie: `warn()` or `report()`) and return `void` to avoid any ambiguity (*I've this `ViolationMessage` object, how do I report it? what do I do with it now?*
6. SLF4J v2 introduced it's own fluent API which may work as reference / inspiration: https://slf4j.org/manual.html#fluent

Reply:
I've created #5039 to track this. I marked the new method as experimental with a disclaimer:

> This will probably never be stabilized, will instead be replaced by a fluent API or something to report violations. Do not use this outside of the PMD codebase. See https://github.com/pmd/pmd/issues/5039